File: android/src/main/java/expo/modules/launchhq/reactnativekeyboardcomposer/KeyboardComposerModule.kt
package expo.modules.launchhq.reactnativekeyboardcomposer

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import android.util.Log

/**
 * Expo module for KeyboardComposer.
 */
class KeyboardComposerModule : Module() {

    override fun definition() = ModuleDefinition {
        Name("KeyboardComposer")

        // KeyboardComposerView
        View(KeyboardComposerView::class) {
            Prop("placeholder") { view: KeyboardComposerView, value: String ->
                view.placeholderText = value
            }

            Prop("text") { view: KeyboardComposerView, value: String ->
                view.text = value
            }

            Prop("minHeight") { view: KeyboardComposerView, value: Float ->
                view.minHeightDp = value
            }

            Prop("maxHeight") { view: KeyboardComposerView, value: Float ->
                view.maxHeightDp = value
            }

            Prop("sendButtonEnabled") { view: KeyboardComposerView, value: Boolean ->
                view.sendButtonEnabled = value
            }

            Prop("showSendButton") { view: KeyboardComposerView, value: Boolean ->
                view.showSendButton = value
            }

            Prop("editable") { view: KeyboardComposerView, value: Boolean ->
                view.editable = value
            }

            Prop("autoFocus") { view: KeyboardComposerView, value: Boolean ->
                view.autoFocus = value
            }

            Prop("blurTrigger") { view: KeyboardComposerView, value: Double ->
                if (value > 0) {
                    view.blur()
                }
            }

            Prop("isStreaming") { view: KeyboardComposerView, value: Boolean ->
                view.isStreaming = value
            }

            Events(
                "onChangeText",
                "onSend",
                "onStop",
                "onHeightChange",
                "onKeyboardHeightChange",
                "onComposerFocus",
                "onComposerBlur"
            )

            // Imperative methods, callable from JS via the component ref.
            // Posted to the view's UI thread because they touch the EditText.
            AsyncFunction("focus") { view: KeyboardComposerView ->
                view.post { view.focus() }
            }

            AsyncFunction("blur") { view: KeyboardComposerView ->
                view.post { view.blur() }
            }

            AsyncFunction("clear") { view: KeyboardComposerView ->
                view.post { view.clear() }
            }
        }

        // KeyboardAwareWrapper
        // Auto-named as "KeyboardComposer_KeyboardAwareWrapper"
        View(KeyboardAwareWrapper::class) {
            Prop("pinToTopEnabled") { view: KeyboardAwareWrapper, value: Boolean ->
                view.pinToTopEnabled = value
            }

            Prop("extraBottomInset") { view: KeyboardAwareWrapper, value: Float ->
                view.extraBottomInset = value
            }

            Prop("scrollToTopTrigger") { view: KeyboardAwareWrapper, value: Double ->
                view.scrollToTopTrigger = value
            }
        }
    }
}

File: ios/KeyboardComposerModule.swift
import ExpoModulesCore

public class KeyboardComposerModule: Module {
  public func definition() -> ModuleDefinition {
    Name("KeyboardComposer")

    // Constants
    Constants([
      "defaultMinHeight": 48.0,
      "defaultMaxHeight": 120.0,
      "contentGap": 0.0 // Gap between last message and composer (points)
    ])

    // Composer View definition
    View(KeyboardComposerView.self) {
      // Props
      Prop("placeholder") { (view: KeyboardComposerView, value: String) in
        view.placeholder = value
      }

      Prop("text") { (view: KeyboardComposerView, value: String) in
        view.text = value
      }

      Prop("minHeight") { (view: KeyboardComposerView, value: CGFloat) in
        view.minHeight = value
      }

      Prop("maxHeight") { (view: KeyboardComposerView, value: CGFloat) in
        view.maxHeight = value
      }

      Prop("sendButtonEnabled") { (view: KeyboardComposerView, value: Bool) in
        view.sendButtonEnabled = value
      }

      Prop("showSendButton") { (view: KeyboardComposerView, value: Bool) in
        view.showSendButton = value
      }

      Prop("editable") { (view: KeyboardComposerView, value: Bool) in
        view.editable = value
      }

      Prop("autoFocus") { (view: KeyboardComposerView, value: Bool) in
        view.autoFocus = value
      }

      Prop("blurTrigger") { (view: KeyboardComposerView, value: Double) in
        if value > 0 {
          view.blur()
        }
      }

      Prop("isStreaming") { (view: KeyboardComposerView, value: Bool) in
        view.isStreaming = value
      }

      Prop("expandedEditorEnabled") { (view: KeyboardComposerView, value: Bool) in
        view.expandedEditorEnabled = value
      }

      Events(
        "onChangeText",
        "onSend",
        "onStop",
        "onHeightChange",
        "onKeyboardHeightChange",
        "onComposerFocus",
        "onComposerBlur"
      )

      // Imperative methods, callable from JS via the component ref.
      // Dispatched to the main thread because they touch UIKit (first responder).
      AsyncFunction("focus") { (view: KeyboardComposerView) in
        DispatchQueue.main.async {
          view.focus()
        }
      }

      AsyncFunction("blur") { (view: KeyboardComposerView) in
        DispatchQueue.main.async {
          view.blur()
        }
      }

      AsyncFunction("clear") { (view: KeyboardComposerView) in
        DispatchQueue.main.async {
          view.clear()
        }
      }
    }

    // Second view in module - keyboard-aware wrapper
    // Auto-named as "KeyboardComposer_KeyboardAwareWrapper"
    View(KeyboardAwareWrapper.self) {
      Prop("pinToTopEnabled") { (view: KeyboardAwareWrapper, value: Bool) in
        view.pinToTopEnabled = value
      }

      Prop("extraBottomInset") { (view: KeyboardAwareWrapper, value: CGFloat) in
        view.extraBottomInset = value
      }

      // No-op on iOS (kept for cross-platform API parity; iOS handles headers via ScrollView contentInset).
      Prop("extraTopInset") { (_: KeyboardAwareWrapper, _: CGFloat) in
        // Intentionally unused
      }
      
      Prop("scrollToTopTrigger") { (view: KeyboardAwareWrapper, value: Double) in
        view.scrollToTopTrigger = value
      }
    }
    
  }
}
